Proposals/NNNN-swift-subprocess.md

```diff
@@ -0,0 +1,882 @@
+    @available(tvOS, unavailable)
+    @available(watchOS, unavailable)
+    public enum TerminationStatus: Sendable, Hashable {
+    #if canImport(WinSDK)
+        public typealias Code = DWORD
+    #else
+        public typealias Code = CInt
+    #endif
+
+    #if canImport(WinSDK)
+        case stillActive
+    #endif
+        case exit(Code)
+        case unhandledException(Code)
+        // A process is terminated successfully when it exited 0
+        public var isSuccess: Bool
+        public var wasUnhandledException: Bool
+    }
+}
+```
+
+
+### `Subprocess.AsyncBytes`
+
+`Subprocess` vends `AsyncBytes` as the concrete implementation of `AsyncSequence<UInt8>`, used by `Subprocess` as the `standardOutput` and `standardError` properties.
+
+```swift
+extension Subprocess {
+    @available(FoundationPreview 0.4, *)
+    @available(iOS, unavailable)
+    @available(tvOS, unavailable)
+    @available(watchOS, unavailable)
+    public struct AsyncBytes: AsyncSequence, Sendable {
+        public typealias Element = UInt8
+        public typealias AsyncIterator = Iterator
+
+        public func makeAsyncIterator() -> Iterator
+
+        @_nonSendable
+        public struct Iterator: AsyncIteratorProtocol {
+            public typealias Element = UInt8
+
+            public mutating func next() async throws -> UInt8?
+        }
+    }
+}
+```
+
+
+## Impact on Existing Code
+
+No impact on existing code is anticipated. All introduced changes are additive.
+
+
+## Future Directions
+
+### Automatic Splitting of `Arguments`
+
+Ideally, the `Arguments` feature should automatically split a string, such as "-a -n 1024 -v 'abc'", into an array of arguments. This enhancement would enable `Arguments` to conform to `ExpressibleByStringLiteral`, allowing developers to conveniently pass either a `String` or `[String]` as `Arguments`.
+
+I decided to defer this feature because it turned out to be a "hard problem" -- different platforms handle arguments differently, requiring careful consideration to ensure correctness.
+
+For reference, Python uses [`shlex.split`](https://docs.python.org/3/library/shlex.html), which could serve as a valuable starting point for implementation.
+
+## Combined `stdout` and `stderr`
+
+In Python's `Subprocess`, developers can merge standard output and standard error into a single stream. This is particularly useful when an executable improperly utilizes standard error as standard output (or vice versa). We should explore the most effective way to achieve this enhancement without introducing confusion to existing parameters—perhaps by introducing a new property.
+
+## Automatic `Subprocess` Cancellation
+
+Currently, `Subprocess` does not terminate the spawned process, even if the `Subprocess` instance goes out of scope. It would be beneficial to investigate whether it is more sensible to attempt terminating the spawned process when the `Subprocess` itself goes out of scope or when the parent task is canceled.
```

Comment:
It really ought to terminate the process. Or else you'll deadlock super easily. Imagine a task group which reads chunks from the sub process's output. Under cancellation that `AsyncSequence` will stop being iterated. That however means that the sub process won't be able to write a lot because the pipe will run full (as the parent is no longer reading). And that means the sub process will never exit --> deadlock.